### DISCUSSION TITLE
Custom Storefront Filtering. Examples of filtering with the new Storefront API?

### DISCUSSION THREAD

**Discussion by DennisKraaijeveld** [Shopify/storefront-api-feedback#72] [Help]
Hi! 

Right now I am using a tag system to filter through my products in each collection but I saw the new Storefront API did add custom filters. Does anyone implemented this already and are there good examples out there? I have no clue yet how to get this to work on each collection.

Using Gatsby for my storefront btw

**Comment by johncraigcole** [comment]
Here's the [examples](https://shopify.dev/api/examples/filter-products) from the Shopify dev docs. Let me know if you have any other questions once you've had a chance to review the provided examples.

**Comment by grichards** [comment]
Yes, but this is such a limited set of attributes to filter by.... how can I filter by Metafield value, which is where all the juicy and useful attributes of my products exist!

**Reply by code-by-gijs** [reply]
You can set it in the "Collection and search filters" on the https://**your-store**.myshopify.com/admin/menus page. Click "add filters" and select the metafield defenitions you would like to filter. If you don't see the "Collection and search filters" section, you will have to update your theme to the latest Shopify version.

**Comment by grichards** [comment]
Thanks @code-by-gijs ... and yes, I am aware of that for use with 2.0 themes, enabling filtration by metafields on collection pages.

However, the question is how to do this via API. I see no examples in the docs about how to construct a GraphQL query that will enable me to get products within a Collection further filtered by a metafield value.

[See here the reference to the possibility in themes.](https://shopify.dev/themes/navigation-search/filtering/storefront-filtering)

[But see here the corresponding API documentation makes no reference to metafields.](https://shopify.dev/api/examples/filter-products)

**Reply by johncraigcole** [reply]
Hey @grichards 👋. As of `2022-04` you can query a product of a collection filtering by a metafield where the type of the metafield are boolean, single_line_text_field, number_integer, number_decimal.

The metafield needs to be listed in the **Collection and search filters** section of Online Store -> Navigation in the admin as an available filter. 

I have a product metafield of type boolean which indicates if a product is used. The following query would allow me to return products of a collection where the used flag is true.

```
query {
	collection(handle: "all-products") {
		id
		title
		products(
			first: 50
			filters: {
				productMetafield: {
					namespace: "bme"
					key: "is_used"
					value: "true"
				}
			}
		) {
			edges {
				cursor
				node {
					id
					title
				}
			}
			pageInfo {
				hasNextPage
				hasPreviousPage
			}
		}
	}
}
```
Hope that helps answer your question.

**Reply by grichards** [reply]
Thank you @johncraigcole . Beautiful! And it does... bookmarking this for future reference. Super awesome Hydrogen site coming soon.

**Reply by johncraigcole** [reply]
Happy to hear that helped. I'll see if I can get a PR up to add a metafield example to the docs. Looking forward to see what you and the developer community do with Hydrogen over the next year. 🎉 🎊

**Reply by grichards** [reply]
Hi @johncraigcole .

FYI, I just tried this today on 2022-07 with my recently imported catalog and I'm getting several errors, here's a summary:

            "message": "Field 'collection' is missing required arguments: id",
            "message": "Field 'collection' doesn't accept argument 'handle'",
            "message": "Field 'products' doesn't accept argument 'filters'",

We're hunting around for solutions... but thought I'd drop the message here just in case.

**Reply by grichards** [reply]
Hello again @johncraigcole 

Ah... my bad, I was attempting this type of query against the Admin API... and I discovered that in the Admin API

- the collection level query would have to be "collectionByHandle"
- that there is no "filters" connection argument for the "products" connection
- there is a "query" connection argument for it, but it does not seem to supply any filtering capability by metafield

**Comment by portedison** [comment]
Just piggy backing on this discussion. I've been implementing the filters on our site (very cool feature). One issue I can't get around easily is reducing the variants when variant (metafields) are applied. I have an example here - https://community.shopify.com/c/shopify-apis-and-sdks/storefront-api-reduce-variants-with-applied-meta-filters/td-p/1708885

**Reply by portedison** [reply]
@johncraigcole Is this something you're working on? We're hoping when we filter down a variant to 'Red' that the images/variants we display match that criteria. The liquid templates handle this case.

**Reply by johncraigcole** [reply]
Hey @portedison I'm assuming the problem here is that the variants connection of the product returned still has all the variants for that product as well? Put another way, even though the product does have a Red Colour variant metafield all other variants associated with that product (blue, green, etc) are also returned.

Assuming I've got the correct understanding, for now, you could ask for the the colour variant metafield in the response which would allow for a filtering of the variants from the query response.

```

query {
	collection(handle: "all-products") {
		id
		title
		products(
			first: 50
			filters: [
				{
					variantMetafield: {
						namespace: "filter"
						key: "colour"
						value: "Silver"
					}
				}
			]
		) {
			edges {
				cursor
				node {
					id
					vendor
					title
					variants(first: 250) {
						edges {
							node {
								title
								metafiledcolour: metafield(namespace: "filter", key: "colour") {
									value
								}
							}
						}
					}
				}
			}
		}
	}
}
```

response would have the first 250 variants for each product and for the variants array you could filter or ignore the colour that doesn't match the filter. 
```
...
"variants": {
	"edges": [
		{
		   "node": {
		      "title": "Silver",
		      "metafieldcolour": {
                 	"value": "Silver"
		       }
                  },
                 {
   		    "node": {
			"title": "Brass",
			"metafieldcolour": {
			   "value": "Brass"
			 }
                  }
        ]
}
```
I'll have to check if this is something already logged. While having the variants returned only be representative of the variant metafield filter specified I believe the approach above could perhaps unblock you in the meantime.

**Reply by portedison** [reply]
Thanks @johncraigcole. That's right. I wondered if that would be the best option. Though I was concerned with rate limits, with 5000+ products. Are there any plans to filter down variants based on variant meta filters in the future?

**Reply by johncraigcole** [reply]
You can only return 250 products at a time, and Shopify has a 100 variant limit per product. Depending on where the Storefront API request is being made (typically client side in the browser) API rate limits for the Storefront API are on the requesting IP. Given a buyers focus would be interacting with filters and you'd likely not be making other Storefront API requests in parallel I don't think rate limits should be an issue. 

As for the bigger question of limiting the variants based on the filter provided this is a great call out. I'll add an issue on our end so that it's tracked. That being said, this feels potentially like a big change to me so I don't want to make any promises on if or when that get's done.
